### ISSUE TITLE
[FEATURE]: Enhancement of Helm deployment status in Garden CLI

### ISSUE THREAD

**Issue by bxsx** [garden-io/garden#6563]
### Background / Motivation

Misconfigurations or missing Kubernetes resources (such as secrets) can cause Garden to timeout with unclear error messages like `context deadline exceeded`. These messages do not pinpoint the actual issue or provide useful guidance for resolution.

### What should the user be able to do?

The user should see the specific issue that caused the timeout, such as `Error: secret "my-service-secret"` not found.

### Why do they want to do this? What problem does it solve?

This feature removes the need to retrieve useful information from the Kubernetes API using external tools (such as `kubectl`, which Garden already uses). It consolidates the relevant information into a single source of truth (the Garden CLI output), which should reduce the time required to resolve issues. Additionally, some clusters, such as CI clusters, may not be accessible to end-users. Without this information available in the CLI output, users may be blocked from taking further action.

### Suggested Implementation(s)

The missing information can be retrieved from K8s events, such as:

```
$ kubectl get events -n "$NAMESPACE" --field-selector involvedObject.kind=Pod
LAST SEEN   TYPE      REASON      OBJECT                                MESSAGE
...
7s          Warning   Failed      pod/my-service-6c12345666-8htrd   Error: secret "my-service-secret" not found
...

# Show failures for the given service
$ kubectl get events -n "$NAMESPACE" --field-selector involvedObject.kind=Pod | grep Failed | grep -E "$(kubectl get pods -n "$NAMESPACE" -l "app.kubernetes.io/name=$SERVICE_NAME" -o jsonpath='{.items[*].metadata.name}')"
14s         Warning   Failed      pod/my-service-6c12345666-8htrd   Error: secret "my-service-secret" not found
```

### How important is this feature for you/your team?

🌵 Not having this feature makes using Garden painful

**Comment by eysi09** [comment]
Thanks for the feature request! 

We are in fact already working on improvements on this front and would love your feedback and suggestions once we the first iteration is ready.

**Comment by eysi09** [comment]
Thanks for the added context! And sorry for the late reply, had a few days off last week. 

> This looks very promising, indeed. How did you manage to get this output? Is it due to log level settings? Unfortunately, I don't think I have ever seen K8s events in the Garden output..

You should always get this if you are using a Deploy action of kind `kubernetes` (and not kind `helm`). Is that not the case for you? In any case I haven't bumped into issues with it myself or seen any reports about it missing. 

But we're currently on adding this output to Deploy actions of kind `helm` and in general making the output more readable and more helpful. Hoping to have a PR up this week.

**Comment by eysi09** [comment]
Irrespective of the above, hence the separate comment, but one other thing I noticed is that when using the `helm` kind we wait for Helm to timeout before exiting with an error. That is basically how Helm works if you use the `--wait` flag (e.g. `helm upgrade --install --wait`). 

This means that if a container won't start or an image pull secret is missing, it won't fail until it timeouts which by default is 300 seconds. 

So I'm also seeing if we can fail fast under those conditions. Unfortunately Helm itself isn't really helpful here but we can use similar heuristics as we do for "normal" K8s manifests although it probably won't work for all cases because a Helm chart can be a lot of things. We might make this opt-in behaviour to begin with since it's technically a change in how we install Helm charts.

**Comment by bxsx** [comment]
> You should always get this if you are using a Deploy action of kind `kubernetes` (and not kind `helm`). Is that not the case for you? In any case I haven't bumped into issues with it myself or seen any reports about it missing.

Good catch! I've just tested it with the kind of `kubernetes` and confirm that I'm getting similar output to what you shared!

> But we're currently on adding this output to Deploy actions of kind `helm` and in general making the output more readable and more helpful. Hoping to have a PR up this week.

Great! I believe this is exactly what I'm looking for. Can't wait to hear any updates on this matter :)

> Irrespective of the above, hence the separate comment, but one other thing I noticed is that when using the helm kind we wait for Helm to timeout before exiting with an error. That is basically how Helm works if you use the --wait flag (e.g. helm upgrade --install --wait).
> This means that if a container won't start or an image pull secret is missing, it won't fail until it timeouts which by default is 300 seconds.

Correct. This is the current behavior, and we're just missing any output that could help in diagnosing the timeouted action.

> So I'm also seeing if we can fail fast under those conditions. Unfortunately Helm itself isn't really helpful here but we can use similar heuristics as we do for "normal" K8s manifests although it probably won't work for all cases because a Helm chart can be a lot of things. We might make this opt-in behavior to begin with since it's technically a change in how we install Helm charts.

I'm not sure if the `helm` kind should support fail fast. I guess, in some scenarios, this might be considered as a breaking change unless it's configured in the manifest file. My understanding is that the `helm` kind shall be considered as a wrapper for Helm, without implicit functional changes in the expected behavior. Anyway, the fail-fast feature is _nice to have_, but the missing outputs from the events are more important (IMHO).

@eysi09 thanks for your detailed replies! I appreciate it!

**Comment by bxsx** [comment]
I'm renaming the issue title to avoid further confusion :)

**Comment by eysi09** [comment]
Just a quick update, we're making good progress and getting this into review.

> I'm not sure if the helm kind should support fail fast. I guess, in some scenarios, this might be considered as a breaking change unless it's configured in the manifest file. 

On closer inspection, turns out that this is what we used to do until recently, on this current Garden major version. 

The reason we changed it was not because of Garden failing fast but rather being "successful fast" and not always waiting for the Helm command to complete after all resources became healthy. We never received any reports about issues related to Garden failing fast, only a certain edge case that came up in the "successful" case and we were a bit overzealous in addressing it by applying `--wait` to everything. 

With the changes I'm making I think we strike a good balance:

- We always use the `--wait` flag but also monitor resources at the same time.
- If the resources are healthy, we always wait for the Helm command to complete. 
- If the resources are unhealthy, we break out early and fail fast. This is what we used to do and we haven't seen issues with it.
- We introduce a config flag to override the fail fast behaviour, but again, we don't expect this to be really needed.
- We always wait for the Helm command to complete if the `atomic` options is set.
